testgen/src/tester.rs

```diff
@@ -1,77 +1,196 @@
+/// with some utility functions operating relative to that directory.
 #[derive(Debug, Clone)]
 pub struct TestEnv {
-    root_dir: String,
-    logs: Vec<String>,
+    /// Directory where the test is being executed
+    current_dir: String,
 }
 
 impl TestEnv {
-    pub fn add_log(&mut self, log: &str) {
-        self.logs.push(log.to_string());
+    pub fn new(current_dir: &str) -> Option<Self> {
+        fs::create_dir_all(current_dir).ok().map(|_| TestEnv {
+            current_dir: current_dir.to_string(),
+        })
     }
 
-    /// Read a file from a path relative to the environment root into a string
-    pub fn read_file(&mut self, path: &str) -> Option<String> {
-        match self.full_path(path) {
-            None => None,
-            Some(full_path) => match fs::read_to_string(&full_path) {
-                Ok(file) => Some(file),
-                Err(_) => None,
-            },
+    pub fn cleanup(&self) -> Option<()> {
+        fs::remove_dir_all(&self.current_dir)
+            .ok()
+            .and(fs::create_dir_all(&self.current_dir).ok())
+    }
+
+    pub fn push(&self, child: &str) -> Option<Self> {
+        let mut path = PathBuf::from(&self.current_dir);
+        path.push(child);
+        path.to_str().and_then(|path| TestEnv::new(path))
+    }
+
+    pub fn current_dir(&self) -> &str {
+        &self.current_dir
+    }
+
+    pub fn logln(&self, msg: &str) -> Option<()> {
+        println!("{}", msg);
+        self.full_path("_log").and_then(|full_path| {
+            fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(full_path)
+                .ok()
+                .and_then(|mut file| file.write_all((String::from(msg) + "\n").as_bytes()).ok())
+        })
+    }
+
+    pub fn logln_to(&self, msg: &str, rel_path: &str) -> Option<()> {
+        println!("{}", msg);
+        self.full_path(rel_path).and_then(|full_path| {
+            fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(full_path)
+                .ok()
+                .and_then(|mut file| file.write_all((String::from(msg) + "\n").as_bytes()).ok())
+        })
+    }
+
+    /// Read a file from a path relative to the environment current dir into a string
+    pub fn read_file(&self, rel_path: &str) -> Option<String> {
+        self.full_path(rel_path)
+            .and_then(|full_path| fs::read_to_string(&full_path).ok())
+    }
+
+    /// Write a file to a path relative to the environment current dir
+    pub fn write_file(&self, rel_path: &str, contents: &str) -> Option<()> {
+        self.full_path(rel_path)
+            .and_then(|full_path| fs::write(full_path, contents).ok())
+    }
+
+    /// Parse a file from a path relative to the environment current dir as the given type
+    pub fn parse_file_as<T: DeserializeOwned>(&self, rel_path: &str) -> Option<T> {
```

Comment:
How about working directly with a path, ie.

```rust
pub fn parse_file_as<T: DeserializeOwned>(&self, rel_path: &Path) -> Option<T>
```

Reply:
changed to `impl AsRef<Path>`